py_plugins/bentoml_deserialization_rce/bentoml_rce_detector.py

```diff
@@ -0,0 +1,206 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A Tsunami plugin for detecting CVE-2024-2912."""
+import pickle
+import time
+from absl import logging
+from google.protobuf import timestamp_pb2
+import tsunami_plugin
+from common.data import network_endpoint_utils
+from common.data import network_service_utils
+from common.net.http.http_client import HttpClient
+from common.net.http.http_headers import HttpHeaders
+from common.net.http.http_request import HttpRequest
+from plugin.payload.payload_generator import PayloadGenerator
+import detection_pb2
+import payload_generator_pb2 as pg
+import plugin_representation_pb2
+import vulnerability_pb2
+
+_VULN_DESCRIPTION = (
+    'The BentoML framework is vulnerable to an insecure deserialization issue that can'
+    ' be exploited by sending a single POST request to any valid endpoint. '
+    'The impact of this is remote code execution.'
+)
+_SLEEP_TIME_SEC = 20
+
+
+class Cve20242912Detector(tsunami_plugin.VulnDetector):
+    """A TsunamiPlugin that detects RCE on the BentoMLtarget."""
+
+    def __init__(
+            self, http_client: HttpClient, payload_generator: PayloadGenerator
+    ):
+        self.http_client = http_client
+        self.payload_generator = payload_generator
+
+    def GetPluginDefinition(self) -> tsunami_plugin.PluginDefinition:
+        """Defines the PluginDefinition for Cve20242912Detector.
+
+    Returns:
+      The PluginDefinition used for the Tsunami engine to identify this plugin.
+    """
+        return tsunami_plugin.PluginDefinition(
+            info=plugin_representation_pb2.PluginInfo(
+                type=plugin_representation_pb2.PluginInfo.VULN_DETECTION,
+                name='Cve20242912VulnDetector',
+                version='1.0',
+                description=_VULN_DESCRIPTION,
+                author='secureness (nosecureness@gmail.com)',
+            )
+        )
+
+    def Detect(
+            self,
+            target: tsunami_plugin.TargetInfo,
+            matched_services: list[tsunami_plugin.NetworkService],
+    ) -> tsunami_plugin.DetectionReportList:
+        """Run detection logic for the BentoML target.
+
+    Args:
+      target: TargetInfo about BentoML Insecure Deserialization.
+      matched_services: A list of network services whose vulnerabilities could
+        be detected by this plugin. "rtsp" for example would be on this list.
```

Comment:
```suggestion
        be detected by this plugin. "ppp" for example would be on this list.
```

With the default port being 3000, an nmap scan that does not recognize the service would give `ppp` as service name.